src/statsmodels/contrasts.jl

```diff
@@ -0,0 +1,233 @@
+    ...
+end
+
+contrasts_matrix(C::MyContrasts, baseind, n) = ...
+```
+
+"""
+abstract AbstractContrasts
+
+## Contrasts + Data = ContrastsMatrix
+type ContrastsMatrix{T}
+    matrix::Matrix{Float64}
+    termnames::Vector{T}
+    levels::Vector{T}
+    contrasts::AbstractContrasts
+end
+
+"""
+    ContrastsMatrix{T}(::AbstractContrasts, levels::Vector{T})
+
+Instantiate contrasts matrix for given data (categorical levels)
+
+If levels are specified in the `AbstractContrasts`, those will be used, and likewise
+for the base level (which defaults to the first level).
+"""
+function ContrastsMatrix{T}(C::AbstractContrasts, lvls::Vector{T})
+
+    ## if levels are defined on C, use those, validating that they line up.
+    ## what does that mean?
+    ##
+    ## C.levels == lvls (best case)
+    ## data levels missing from contrast: would generate empty/undefined rows. 
+    ## better to filter data frame first
+    ## contrast levels missing from data: would have empty columns, generate a
+    ## rank-deficient model matrix.
+    c_lvls = convert(typeof(lvls), get(C.levels, lvls))
+    mismatched_lvls = symdiff(c_lvls, lvls)
+    isempty(mismatched_lvls) || error("Contrasts levels not found in data or vice-versa: ", mismatched_lvls)
+
+    n = length(c_lvls)
+    n > 1 || error("not enough degrees of freedom to define contrasts")
+    
+    ## find index of base level. use C.base, then default (1).
+    baseind = isnull(C.base) ? 1 : findfirst(c_lvls, convert(eltype(lvls), get(C.base)))
+    baseind > 0 || error("Base level $(C.base) not found in levels")
+    
+    not_base = [1:(baseind-1); (baseind+1):n]
+    tnames = c_lvls[not_base]
+
+    mat = contrasts_matrix(C, baseind, n)
+
+    ContrastsMatrix(mat, tnames, c_lvls, C)
+end
+
+ContrastsMatrix(C::AbstractContrasts, v::PooledDataArray) = ContrastsMatrix(C, levels(v))
+
+
+nullify(x::Nullable) = x
+nullify(x) = Nullable(x)
+
+## Making a contrast type T only requires that there be a method for
+## contrasts_matrix(T, v::PooledDataArray). The rest is boilerplate.
+##
+for contrastType in [:TreatmentContrasts, :SumContrasts, :HelmertContrasts]
+    @eval begin
+        type $contrastType <: AbstractContrasts
+            base::Nullable{Any}
+            levels::Nullable{Vector}
+        end
+        ## constructor with optional keyword arguments, defaulting to Nullables
+        $contrastType(;
+                      base=Nullable{Any}(),
+                      levels=Nullable{Vector}()) = 
+                          $contrastType(nullify(base),
+                                        nullify(levels))
+    end
+end
+
+"""
+    DummyContrasts
+
+One indicator (1 or 0) column for each level, __including__ the base level.
+
+Needed internally when a term is non-redundant with lower-order terms (e.g., in
+~0+x vs. ~1+x, or in the interactions terms in ~1+x+x&y vs. ~1+x+y+x&y. In the
+non-redundant cases, we can (probably) expand x into length(levels(x)) columns
```

Comment:
"probably" is a bit worrying here. I'd remove it, the parenthesis at the end is explicit enough. But maybe you could be more specific about what "something dumb" could be, and what could be the consequences? We don't want to return misleading results, or even let users believe this can be the case.

Also add bacticks around formulas and `length(levels(x))`.